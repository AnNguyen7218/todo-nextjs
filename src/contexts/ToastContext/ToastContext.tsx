import React, { useContext } from 'react';

type ContextProps = {
  toastMessage: (message: string, toastTheme: ToastTheme) => void;
};

export const ToastContext = React.createContext<Partial<ContextProps>>({});
export const useToastContext = () => useContext(ToastContext);
