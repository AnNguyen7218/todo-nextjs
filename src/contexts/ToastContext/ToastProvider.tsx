import React, { useState } from 'react';
import { toast } from 'react-toastify';
import { Toast, ToastMessage } from 'components';
import { ToastContext } from './ToastContext';

type Props = {
  children: React.ReactChild;
};

const ToastProvider: React.FC<Props> = ({ children }) => {
  const [theme, setTheme] = useState<ToastTheme>('success');
  const toastMessage = (
    message: string,
    toastTheme: ToastTheme = 'success'
  ) => {
    setTheme(toastTheme);
    toast(
      ({ closeToast }: { closeToast: () => void }) => (
        <ToastMessage
          message={message}
          closeToast={closeToast}
          theme={toastTheme}
        />
      ),
      {
        position: 'top-center',
      }
    );
  };

  return (
    <ToastContext.Provider value={{ toastMessage }}>
      <Toast theme={theme} />
      {children}
    </ToastContext.Provider>
  );
};

export default ToastProvider;
