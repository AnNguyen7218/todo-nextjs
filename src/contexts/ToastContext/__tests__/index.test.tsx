import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import ToastProvider from '../ToastProvider';
import { ToastContext } from '../ToastContext';

afterEach(() => {
  cleanup();
  jest.useFakeTimers();
});

describe('Toast context', () => {
  test('Toast context should match snapshot', () => {
    const { container } = render(
      <ToastProvider>
        <div>Test for toast</div>
      </ToastProvider>
    );
    expect(container).toMatchSnapshot();
  });

  describe('Toast context should show', () => {
    test('Toast show success message', () => {
      const { asFragment } = render(
        <ToastProvider>
          <ToastContext.Consumer>
            {({ toastMessage }) => {
              toastMessage && toastMessage('Success', 'success');
              return <div>Test for toast</div>;
            }}
          </ToastContext.Consumer>
        </ToastProvider>
      );
      jest.runAllTimers();
      expect(asFragment()).toMatchSnapshot();
    });

    test('Toast error message', () => {
      const { asFragment } = render(
        <ToastProvider>
          <ToastContext.Consumer>
            {({ toastMessage }) => {
              toastMessage && toastMessage('Error', 'error');
              return <div>Test for toast</div>;
            }}
          </ToastContext.Consumer>
        </ToastProvider>
      );
      jest.runAllTimers();
      expect(asFragment()).toMatchSnapshot();
    });

    test('Toast warning message', () => {
      const { asFragment } = render(
        <ToastProvider>
          <ToastContext.Consumer>
            {({ toastMessage }) => {
              toastMessage && toastMessage('Warning', 'warning');
              return <div>Test for toast</div>;
            }}
          </ToastContext.Consumer>
        </ToastProvider>
      );
      jest.runAllTimers();
      expect(asFragment()).toMatchSnapshot();
    });

    test('Toast close message', () => {
      const { asFragment, getByTestId } = render(
        <ToastProvider>
          <ToastContext.Consumer>
            {({ toastMessage }) => {
              toastMessage && toastMessage('Warning', 'warning');
              return <div>Test for toast</div>;
            }}
          </ToastContext.Consumer>
        </ToastProvider>
      );
      jest.runAllTimers();
      expect(asFragment()).toMatchSnapshot();
      const close = getByTestId('close-icon');
      expect(close).not.toBeNull();
      fireEvent.click(close);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
