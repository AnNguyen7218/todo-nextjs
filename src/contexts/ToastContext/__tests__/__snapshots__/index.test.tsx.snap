// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast context Toast context should match snapshot 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div>
    Test for toast
  </div>
</div>
`;

exports[`Toast context Toast context should show Toast close message 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  >
    <div
      class="Toastify__toast-container Toastify__toast-container--top-center toast-container"
    >
      <div
        class="Toastify__toast Toastify__toast--default toast toast--warning"
        style=""
      >
        <div
          class="Toastify__toast-body toast-body"
          role="alert"
        >
          <div
            class="container"
          >
            <ic-warning.svg
              classname="icon color--warning"
            />
            <p
              class="message color--warning"
            >
              Warning
            </p>
            <ic-close.svg
              classname="icon color--warning"
              data-testid="close-icon"
            />
          </div>
        </div>
        <div
          class="Toastify__progress-bar Toastify__progress-bar--animated Toastify__progress-bar--default"
          style="animation-duration: 5000ms; animation-play-state: running; opacity: 0;"
        />
      </div>
    </div>
  </div>
  <div>
    Test for toast
  </div>
</DocumentFragment>
`;

exports[`Toast context Toast context should show Toast close message 2`] = `
<DocumentFragment>
  <div
    class="Toastify"
  >
    <div
      class="Toastify__toast-container Toastify__toast-container--top-center toast-container"
      style="pointer-events: none;"
    >
      <div
        class="Toastify__toast Toastify__toast--default toast toast--warning Toastify__slide-exit--top-center"
        style="animation-fill-mode: forwards; animation-duration: 0.75s;"
      >
        <div
          class="Toastify__toast-body toast-body"
        >
          <div
            class="container"
          >
            <ic-warning.svg
              classname="icon color--warning"
            />
            <p
              class="message color--warning"
            >
              Warning
            </p>
            <ic-close.svg
              classname="icon color--warning"
              data-testid="close-icon"
            />
          </div>
        </div>
        <div
          class="Toastify__progress-bar Toastify__progress-bar--animated Toastify__progress-bar--default"
          style="animation-duration: 5000ms; animation-play-state: running; opacity: 0;"
        />
      </div>
    </div>
  </div>
  <div>
    Test for toast
  </div>
</DocumentFragment>
`;

exports[`Toast context Toast context should show Toast error message 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  >
    <div
      class="Toastify__toast-container Toastify__toast-container--top-center toast-container"
    >
      <div
        class="Toastify__toast Toastify__toast--default toast toast--error"
        style=""
      >
        <div
          class="Toastify__toast-body toast-body"
          role="alert"
        >
          <div
            class="container"
          >
            <ic-error-circle.svg
              classname="icon color--error"
            />
            <p
              class="message color--error"
            >
              Error
            </p>
            <ic-close.svg
              classname="icon color--error"
              data-testid="close-icon"
            />
          </div>
        </div>
        <div
          class="Toastify__progress-bar Toastify__progress-bar--animated Toastify__progress-bar--default"
          style="animation-duration: 5000ms; animation-play-state: running; opacity: 0;"
        />
      </div>
    </div>
  </div>
  <div>
    Test for toast
  </div>
</DocumentFragment>
`;

exports[`Toast context Toast context should show Toast show success message 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  >
    <div
      class="Toastify__toast-container Toastify__toast-container--top-center toast-container"
    >
      <div
        class="Toastify__toast Toastify__toast--default toast toast--success"
        style=""
      >
        <div
          class="Toastify__toast-body toast-body"
          role="alert"
        >
          <div
            class="container"
          >
            <ic-check-circle.svg
              classname="icon color--success"
            />
            <p
              class="message color--success"
            >
              Success
            </p>
            <ic-close.svg
              classname="icon color--success"
              data-testid="close-icon"
            />
          </div>
        </div>
        <div
          class="Toastify__progress-bar Toastify__progress-bar--animated Toastify__progress-bar--default"
          style="animation-duration: 5000ms; animation-play-state: running; opacity: 0;"
        />
      </div>
    </div>
  </div>
  <div>
    Test for toast
  </div>
</DocumentFragment>
`;

exports[`Toast context Toast context should show Toast warning message 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  >
    <div
      class="Toastify__toast-container Toastify__toast-container--top-center toast-container"
    >
      <div
        class="Toastify__toast Toastify__toast--default toast toast--warning"
        style=""
      >
        <div
          class="Toastify__toast-body toast-body"
          role="alert"
        >
          <div
            class="container"
          >
            <ic-warning.svg
              classname="icon color--warning"
            />
            <p
              class="message color--warning"
            >
              Warning
            </p>
            <ic-close.svg
              classname="icon color--warning"
              data-testid="close-icon"
            />
          </div>
        </div>
        <div
          class="Toastify__progress-bar Toastify__progress-bar--animated Toastify__progress-bar--default"
          style="animation-duration: 5000ms; animation-play-state: running; opacity: 0;"
        />
      </div>
    </div>
  </div>
  <div>
    Test for toast
  </div>
</DocumentFragment>
`;
